### DISCUSSION TITLE
Better extends (layers)

### DISCUSSION THREAD

**Discussion by localusercamp** [nuxt/nuxt#20380] [Ideas]
I think that [layers](https://nuxt.com/docs/getting-started/layers) is awesome feature and could have more use-cases and options then just extending from another projects.

`extends` could be used not only for extending another projects or themes but for composing your own project. In the root folder of nuxt app you could create `layers` directory. Then you could create `layers/auth` and `layers/posts` directories which contains pages, components and so on. This would be very helpful for implementing modular architecture. Here a few **ideas** of how it could be done:
- Allow users not to create `nuxt.config.*` in every layer. For now every layer will contain config file with `defineNuxtConfig({})`.
- Provide more control on `extends` behaviour. 
  - Allow [modules](https://nuxt.com/docs/guide/going-further/modules/#module-author-guide) to modify `extends` config option as all other options through `nuxt.options.extends`; 
  - Allow users to define something like `layer` config option in which all nuxt app folders might be specified. This might be helpful to restrict what exactly will be inherited from your project. For now every composable and page will be inherited through `extends` which might be unwanted behaviour in some cases.
    ```typescript
    defineNuxtConfig({
      layer: {
        expose: ['composables', 'utils', 'layouts']
      }
    })
    ```
  - Allow specify layer as `scoped`. Scoped layers expose only what was specified in `layer.expose`. It means that all auto-imports and components which was used in layer will not be availalbe to "extenders". This will allow to aviod name collisions and restrict unwanted  layer usage from another layers.

I think if any of above will be implemented it would be a great step further into more customizable **nuxt application templates**.
This features combined with modules could form a great application templates for different architectures such as DDD and modular architecture.

I'll be glad to discuss this. 👌

**Comment by davestewart** [comment]
Agree with modules being able to define or modify layers.

I submitted that as an idea [here](https://github.com/nuxt/nuxt/discussions/26837) with some background and initial discoveries.

I like the expose idea too; saves having to manually configure.

I think having a config file per layer is fine. FWIW I use a helper called `defineLayerConfig()` in which I set up components, content, route rules, etc, so a layer config looks something like this:

```ts
import { defineLayerConfig } from '../base/utils/nuxt'

export default defineLayerConfig('home', {
  prerender: true,
  components: true,
  content: {
    prefix: false
  },
})
```

Regarding your auto-layers idea, there's nothing to stop you writing a small helper to glob said layers, and build the `extends` in the root `nuxt.config.ts` file.
